File: frontend/src/lib/compare.ts
import { api } from "./api";
import type { MediaType } from "./media";
import type { WatchStatus } from "./library";

export type LibrarySignal = {
  status: WatchStatus | null;
  is_favorite: boolean;
  rating: number | null;
};

export type CommonTitle = {
  tmdb_id: number;
  media_type: MediaType;
  title: string;
  poster_url: string | null;
  you: LibrarySignal;
  friend: LibrarySignal;
};

export type RecommendedTitle = {
  tmdb_id: number;
  media_type: MediaType;
  title: string;
  poster_url: string | null;
  friend_is_favorite: boolean;
  friend_rating: number | null;
  friend_status: WatchStatus | null;
};

export type CompareResponse = {
  friend: { id: string; username: string };
  compatibility_score: number;
  common_count: number;
  total_count: number;
  common_titles: CommonTitle[];
  recommendations: RecommendedTitle[];
};

export async function getCompatibility(friendId: string): Promise<CompareResponse> {
  const { data } = await api.get<CompareResponse>(`/friends/${friendId}/compare`);
  return data;
}

File: backend/app/schemas/compare.py
from typing import Literal

from pydantic import BaseModel

from app.schemas.friends import FriendUserOut


class LibrarySignal(BaseModel):
    status: str | None
    is_favorite: bool
    rating: int | None


class CommonTitle(BaseModel):
    tmdb_id: int
    media_type: Literal["movie", "tv"]
    title: str
    poster_url: str | None
    you: LibrarySignal
    friend: LibrarySignal


class RecommendedTitle(BaseModel):
    tmdb_id: int
    media_type: Literal["movie", "tv"]
    title: str
    poster_url: str | None
    friend_is_favorite: bool
    friend_rating: int | None
    friend_status: str | None


class CompareResponse(BaseModel):
    friend: FriendUserOut
    compatibility_score: float
    common_count: int
    total_count: int
    common_titles: list[CommonTitle]
    recommendations: list[RecommendedTitle]
